astropy/coordinates/builtin_frames/utils.py

```diff
@@ -34,29 +35,18 @@
 _DEFAULT_PM = (0.035, 0.29)*u.arcsec
 
 
-def cartrepr_from_matmul(pmat, coo, transpose=False):
-    """
-    Note that pmat should be an ndarray, *not* a matrix.
+def matrix_product(*args):
+    """Matrix multiply all arguments together.
+
+    Arguments should have dimension 2 or larger. Larger dimensional objects
+    are interpreted as stacks of matrices residing in the last two dimensions.
     """
-    if pmat.shape[-2:] != (3, 3):
-        raise ValueError("tried to do matrix multiplication with an array that "
-                         "doesn't end in 3x3")
-    if coo.isscalar and pmat.shape == (3, 3):
-        # a simpler path for scalar coordinates
-        if transpose:
-            pmat = pmat.T
-        newxyz = np.sum(pmat * coo.cartesian.xyz, axis=-1)
-    else:
-        xyz = coo.cartesian.xyz.T
-        # these expression are the same as iterating over the first dimension of
-        # pmat and xyz and doing matrix multiplication on each in turn.  resulting
-        # dimension is <coo shape> x 3
-        pmat = pmat.reshape(pmat.size//9, 3, 3)
-        if transpose:
-            pmat = pmat.transpose(0, 2, 1)
-        newxyz = np.sum(pmat * xyz.reshape(xyz.size//3, 1, 3), axis=-1).T
-
-    return CartesianRepresentation(newxyz)
+    return reduce(matmul, args)
+
+
+def matrix_transpose(m):
+    """Transpose a matrix or stack of matrices by swapping the last two axes."""
```

Comment:
Also done. Seeing it again, I'm wondering slightly about just doing this explicitly... But maybe readability really wins here.  It will certainly not matter much for speed.